Fix the FTR XBlock for Hawthorn

OmarIthawi opened the pull request:
The following error occurs when using this XBlock on Hawthorn.


### TODO

 - [x] Test also in Ficus (I don't have Ginkgo :man_shrugging:)

### Questions

 - [ ] How do I get the student responses?


### Before
<kbd>![haw](https://user-images.githubusercontent.com/645156/48679149-2b112b00-eb95-11e8-86ab-f32b99a0f7f0.png)</kbd>

### After

This PR fixes it:

<kbd>![haw-fixed](https://user-images.githubusercontent.com/645156/48679153-382e1a00-eb95-11e8-9f1b-55e7b58b1247.png)</kbd>

OmarIthawi committed:
Fix the FTR XBlock for Hawthorn

OmarIthawi commented:
@caesar2164 Could you please take a look?

caesar2164 commented:
Thanks for the PR @OmarIthawi!

I'll be looking at it this sprint (next two weeks).

OmarIthawi commented:
Thanks @caesar2164! Looking forward for your review :slightly_smiling_face:

aucanpool commented:
How can I add this commit in my requirements/edx/github.text file?

OmarIthawi commented:
@aucanpool you can use this branch: https://github.com/OmarIthawi/xblock-free-text-response/tree/omar/hawthorn

aucanpool commented:
Thank you very much @OmarIthawi. I am using your branch.

caesar2164 commented on freetextresponse/tests.py:
```diff
@@ -11,9 +11,10 @@
 
 from django.db import IntegrityError
-from django.template.loader import get_template
+from django.template.context import Context
```
Why the switch from `get_template()` to `Context()`?

stvstnfrd replied:
@caesar2164 I don't think it's switch as much as it's just a deletion of one import (`get_template`) and the addition of another (`Context`).

OmarIthawi replied:
@caesar2164 @stvstnfrd I've used `render_django_template` which is more XBlock friendly, especially for `i18n`.

caesar2164 commented on freetextresponse/freetextresponse.py:
```diff
@@ -297,10 +291,15 @@ def student_view(self, context={}):
+            context=Context(context),
+            i18n_service=self.runtime.service(self, "i18n"),
+        )
+
```
let's get rid of extra blank lines

OmarIthawi replied:
Done

caesar2164 commented on freetextresponse/freetextresponse.py:
```diff
@@ -26,16 +25,14 @@
-    StudioEditableXBlockMixin,
-    XBlock,
-):
+class FreeTextResponse(EnforceDueDates, MissingDataFetcherMixin, StudioEditableXBlockMixin, XBlock):
```
please keep the arguments in the multiline format (especially with the trailing comma at the end), it makes diffs cleaner and reduces merge conflicts...

stvstnfrd replied:
Specifically here, it doesn't look like anything changed except reformatting the lines?
Especially in that case, we should not make this change as it's not related to the issue at hand.

OmarIthawi replied:
@caesar2164 @stvstnfrd I've used `render_django_template` which is more XBlock friendly, especially for `i18n`.

caesar2164 reviewed (changes requested):
Thanks @OmarIthawi! 

This is looking really good. I just have a few comments/questions.

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -297,10 +291,15 @@ def student_view(self, context={}):
             }
         )
-        template = get_template('freetextresponse_view.html')
+
```
Yeah, small style nitpick, but lets remove this empty line here.

stvstnfrd commented on freetextresponse/tests.py:
```diff
@@ -56,7 +57,12 @@ def make_an_xblock(cls, **kw):
+        runtime = Mock(
+            course_id=course_id,
+            service=Mock(
+                return_value=Mock(_catalog={}),
```
What is this used by?

OmarIthawi replied:
So the `i18n` service doesn't freak out. I've added an inline comment to clarify.

stvstnfrd replied:
Thanks @OmarIthawi

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -297,10 +291,15 @@ def student_view(self, context={}):
+        template = self.loader.render_django_template(
+            'public/templates/freetextresponse_view.html',
+            context=Context(context),
+            i18n_service=self.runtime.service(self, "i18n"),
```
Little nitpick, but can you change this to use single quotes? `'i18n'`

OmarIthawi replied:
Done

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -2,10 +2,8 @@
 This is the core logic for the Free-text Response XBlock
 """
 from enum import Enum
-import pkg_resources
```
This was just unused?

OmarIthawi replied:
Apparently, yeah.

stvstnfrd commented on requirements.txt:
```diff
@@ -1,6 +1,6 @@
 django<1.9,>=1.8
 git+https://github.com/edx/XBlock.git#egg=XBlock
-git+https://github.com/edx/xblock-utils.git@v1.0.0#egg=xblock-utils==v1.0.0
+git+https://github.com/edx/xblock-utils.git@v1.1.1#egg=xblock-utils==v1.1.1
```
What is the impact of this re: ginkgo vs hawthorn, etc.?

OmarIthawi replied:
Thanks @stvstnfrd. I'll test this one, adding it to the pull request TODO list.

OmarIthawi replied:
BTW @stvstnfrd, I've already tested `xblock-utils.git==v1.1.1` on Ficus, and it worked fine. However, I'm happy to re-test in Ficus specifically for this XBlock if you think it's required.

As for Ginkgo, I have no Ginkgo sandbox available to test on, so would appreciate your help.

stvstnfrd replied:
Thanks for testing @OmarIthawi . I looked at the relevant commits and this does look like it should be backwards compatible.

stvstnfrd commented on tox.ini:
```diff
@@ -13,7 +13,7 @@ deps =
     -rrequirements.txt
     pep8
-commands = {envbindir}/pep8 freetextresponse/
+commands = {envbindir}/pep8 --max-line-length=120 freetextresponse/
```
Which lines necessitated this?

OmarIthawi replied:
I must have forgot it. Thanks @stvstnfrd!

stvstnfrd reviewed (changes requested):
Thanks for taking this on!

I have a few questions and a few style nitpicks, but otherwise this is looking good.
I think we'll also need to bump the version number; should be in `setup.py`, I think?

Note: I haven't had a chance to spin this up in a devstack yet, so give me a chance to do that still.

Thanks again and let me know if you have questions/concerns with the feedback!

stvstnfrd commented on freetextresponse/freetextresponse.py:
```diff
@@ -297,10 +291,15 @@ def student_view(self, context={}):
-        template = get_template('freetextresponse_view.html')
+
+        template = self.loader.render_django_template(
+            'public/templates/freetextresponse_view.html',
```
Wait, I didn't think the templates were in the `public` directory...

I thought `templates` was a sibling of `public`, isn't it?
Related, do we also need to update `setup.py` to ensure the `templates` directory is packaged,
similar to https://github.com/Stanford-Online/xblock-submit-and-compare/pull/25

OmarIthawi replied:
Oops, moving that out. Thanks @stvstnfrd.

OmarIthawi commented:
@caesar2164 @stvstnfrd Thank you for the review! Will get back to the review soon!

OmarIthawi commented:
@caesar2164 @stvstnfrd Thank you for review, I've addressed all of the comments.

I still have to test the updated version in both Hawthorn (just another check) and Ficus.

Please please a look and let me know what do you think.

OmarIthawi commented:
Good news, it works on both Ficus and Hawthorn. Ticking those boxes in the PR review.

@caesar2164 @stvstnfrd Please take another look.

### Ficus
<kbd>![image](https://user-images.githubusercontent.com/645156/49695019-2096ef80-fb9d-11e8-8410-3701f8edc220.png)</kbd>

### Hawthorn
<kbd>![image](https://user-images.githubusercontent.com/645156/49696228-c4899680-fbaf-11e8-9f73-0087084df4fa.png)</kbd>

caesar2164 commented on freetextresponse/freetextresponse.py:
```diff
@@ -298,10 +297,13 @@ def student_view(self, context={}):
+        template = self.loader.render_django_template(
+            'templates/freetextresponse_view.html',
+            context=Context(context),
+            i18n_service=self.runtime.service(self, 'i18n'),
```
Hmm, I'm getting an error on our Ginkgo based devstack:

![screen shot 2018-12-10 at 11 50 35](https://user-images.githubusercontent.com/3364609/49757669-d59adc00-fc71-11e8-9b3b-372ff321e34f.png)

What version of xblock-utils are you running in your Ficus devstack? Our Ginkgo stacks are running v1.0.5, and that argument seems to have been added in xblock-utils v1.1.0? (https://github.com/edx/xblock-utils/commit/ba54084a572a618b9248236403c1200f32d4f889)

OmarIthawi replied:
@caesar2164 Looks like I already have `xblock-utils==1.1.1` installed already. We have a couple of other PRs needing i18n and therefore using `xblock-utils==1.1.1`.

How do you suggest moving on with this issue? Should we require updating the xblock-utils or doing a nasty patch to avoid breaking the XBlock on pre-Hawthorn releases.

stvstnfrd replied:
@caesar2164 Are you still getting this issue or has it been resolved?

caesar2164 replied:
I believe so, looks like @OmarIthawi already pushed the version of xblock-utils to 1.1.1

OmarIthawi commented:
@caesar2164 I think this PR is ready for merging. 

If upgrading `xblock-utils` is not an option for your fork, there's a workaround, or should I say a hack, in which we can avoid it.

caesar2164 commented:
@OmarIthawi, sorry for the hiatus, I forgot to mention that Stanford takes a 2 week "closure" in winter.

On my end, this seems good to go. Only @stvstnfrd's review left.

OmarIthawi commented:
Thanks @caesar2164! No worries at all, closure is always good :smiley:

OmarIthawi commented:
@stvstnfrd could you please take a look and let me know what do you think?

cc: @caesar2164

caesar2164 commented:
Thanks for the PR @OmarIthawi! Merged!

OmarIthawi commented:
@caesar2164 @stvstnfrd Thank you for creating this XBlock and for your help here!
